File: go/internal/e2e/rpc_shell_and_fleet_e2e_test.go
package e2e

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

// Mirrors dotnet/test/RpcShellAndFleetTests.cs (snapshot category "rpc_shell_and_fleet").
func TestRpcShellAndFleetE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should execute shell command", func(t *testing.T) {
		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		markerPath := filepath.Join(ctx.WorkDir, "shell-rpc-"+randomHex(t)+".txt")
		const marker = "copilot-sdk-shell-rpc"

		cwd := ctx.WorkDir
		result, err := session.RPC.Shell.Exec(t.Context(), &rpc.ShellExecRequest{
			Command: writeFileCommand(markerPath, marker),
			Cwd:     &cwd,
		})
		if err != nil {
			t.Fatalf("Failed to call session.shell.exec: %v", err)
		}
		if strings.TrimSpace(result.ProcessID) == "" {
			t.Fatal("Expected non-empty processId from shell.exec")
		}

		waitForFileText(t, markerPath, marker)
	})

	t.Run("should kill shell process", func(t *testing.T) {
		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		var command string
		if runtime.GOOS == "windows" {
			command = "powershell -NoLogo -NoProfile -Command \"Start-Sleep -Seconds 30\""
		} else {
			command = "sleep 30"
		}

		exec, err := session.RPC.Shell.Exec(t.Context(), &rpc.ShellExecRequest{Command: command})
		if err != nil {
			t.Fatalf("Failed to call session.shell.exec: %v", err)
		}
		if strings.TrimSpace(exec.ProcessID) == "" {
			t.Fatal("Expected non-empty processId from shell.exec")
		}

		kill, err := session.RPC.Shell.Kill(t.Context(), &rpc.ShellKillRequest{ProcessID: exec.ProcessID})
		if err != nil {
			t.Fatalf("Failed to call session.shell.kill: %v", err)
		}
		if !kill.Killed {
			t.Errorf("Expected shell.kill to report Killed=true, got %+v", kill)
		}
	})

	t.Run("should start fleet and complete custom tool task", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		markerPath := filepath.Join(ctx.WorkDir, "fleet-rpc-"+randomHex(t)+".txt")
		const marker = "copilot-sdk-fleet-rpc"
		const toolName = "record_fleet_completion"

		type RecordParams struct {
			Content string `json:"content" jsonschema:"Content to record"`
		}
		recordTool := copilot.DefineTool(toolName, "Records completion of the fleet validation task.",
			func(params RecordParams, inv copilot.ToolInvocation) (string, error) {
				if err := os.WriteFile(markerPath, []byte(params.Content), 0644); err != nil {
					return "", err
				}
				return params.Content, nil
			})

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools:               []copilot.Tool{recordTool},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		prompt := fmt.Sprintf("Use the %s tool with content '%s', then report that the fleet task is complete.", toolName, marker)
		promptCopy := prompt

		fleet, err := session.RPC.Fleet.Start(t.Context(), &rpc.FleetStartRequest{Prompt: &promptCopy})
		if err != nil {
			t.Fatalf("Failed to call session.fleet.start: %v", err)
		}
		if !fleet.Started {
			t.Fatal("Expected fleet.start to report Started=true")
		}

		waitForFileText(t, markerPath, marker)

		// Fleet-mode tasks do not emit SessionIdleEvent; poll session messages until the
		// assistant reply contains the expected text.
		messages := waitForFleetCompletion(t, session, "fleet task")

		var sawUser, sawAssistant bool
		var sawToolStart, sawToolComplete bool
		for _, evt := range messages {
			switch d := evt.Data.(type) {
			case *copilot.UserMessageData:
				if strings.Contains(d.Content, prompt) {
					sawUser = true
				}
			case *copilot.AssistantMessageData:
				if strings.Contains(strings.ToLower(d.Content), "fleet task") {
					sawAssistant = true
				}
			case *copilot.ToolExecutionStartData:
				if d.ToolName == toolName {
					sawToolStart = true
				}
			case *copilot.ToolExecutionCompleteData:
				if d.Success && d.Result != nil && strings.Contains(d.Result.Content, marker) {
					sawToolComplete = true
				}
			}
		}

		if !sawUser {
			t.Errorf("Expected user message containing original prompt; messages: %d", len(messages))
		}
		if !sawAssistant {
			t.Errorf("Expected assistant message containing 'fleet task'")
		}
		if !sawToolStart {
			t.Errorf("Expected ToolExecutionStart for %q", toolName)
		}
		if !sawToolComplete {
			t.Errorf("Expected successful ToolExecutionComplete with content containing %q", marker)
		}
	})
}

func randomHex(t *testing.T) string {
	t.Helper()
	var buf [8]byte
	if _, err := rand.Read(buf[:]); err != nil {
		t.Fatalf("Failed to generate random bytes: %v", err)
	}
	return hex.EncodeToString(buf[:])
}

func writeFileCommand(markerPath, marker string) string {
	if runtime.GOOS == "windows" {
		return fmt.Sprintf("powershell -NoLogo -NoProfile -Command \"Set-Content -LiteralPath '%s' -Value '%s'\"", markerPath, marker)
	}
	return fmt.Sprintf("sh -c \"printf '%%s' '%s' > '%s'\"", marker, markerPath)
}

func waitForFileText(t *testing.T, path, expected string) {
	t.Helper()
	deadline := time.Now().Add(30 * time.Second)
	for time.Now().Before(deadline) {
		if data, err := os.ReadFile(path); err == nil && strings.Contains(string(data), expected) {
			return
		}
		time.Sleep(100 * time.Millisecond)
	}
	t.Fatalf("Timed out waiting for shell command to write %q to %q", expected, path)
}

func waitForFleetCompletion(t *testing.T, session *copilot.Session, contentNeedle string) []copilot.SessionEvent {
	t.Helper()
	deadline := time.Now().Add(120 * time.Second)
	for time.Now().Before(deadline) {
		messages, err := session.GetMessages(t.Context())
		if err == nil {
			for _, evt := range messages {
				if d, ok := evt.Data.(*copilot.AssistantMessageData); ok && strings.Contains(strings.ToLower(d.Content), contentNeedle) {
					return messages
				}
			}
		}
		time.Sleep(250 * time.Millisecond)
	}
	t.Fatal("Timed out waiting for fleet-mode assistant reply")
	return nil
}

File: nodejs/test/e2e/rpc_shell_and_fleet.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { describe, expect, it } from "vitest";
import { z } from "zod";
import { approveAll, defineTool } from "../../src/index.js";
import type { CopilotSession, SessionEvent } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Shell and fleet RPC", async () => {
    const { copilotClient: client, workDir } = await createSdkTestContext();

    function createWriteFileCommand(markerPath: string, marker: string): string {
        if (os.platform() === "win32") {
            return `powershell -NoLogo -NoProfile -Command "Set-Content -LiteralPath '${markerPath}' -Value '${marker}'"`;
        }
        return `sh -c "printf '%s' '${marker}' > '${markerPath}'"`;
    }

    async function waitForFileText(
        filePath: string,
        expected: string,
        timeoutMs = 30_000
    ): Promise<void> {
        const deadline = Date.now() + timeoutMs;
        while (Date.now() < deadline) {
            if (fs.existsSync(filePath)) {
                const content = fs.readFileSync(filePath, "utf8");
                if (content.includes(expected)) {
                    return;
                }
            }
            await new Promise((resolve) => setTimeout(resolve, 100));
        }
        throw new Error(
            `Timed out waiting for shell command to write '${expected}' to '${filePath}'.`
        );
    }

    async function waitForMessages(
        session: CopilotSession,
        predicate: (events: SessionEvent[]) => boolean,
        timeoutMs = 120_000
    ): Promise<SessionEvent[]> {
        // Fleet-mode tasks do not emit session.idle on completion, so polling the
        // session message list is the simplest way to wait for a satisfying state.
        const deadline = Date.now() + timeoutMs;
        while (Date.now() < deadline) {
            const messages = await session.getMessages();
            if (predicate(messages)) {
                return messages;
            }
            await new Promise((resolve) => setTimeout(resolve, 250));
        }
        throw new Error("Timed out waiting for fleet-mode assistant reply to satisfy predicate.");
    }

    it("should execute shell command", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        const markerPath = path.join(
            workDir,
            `shell-rpc-${Date.now()}-${Math.random().toString(36).slice(2)}.txt`
        );
        const marker = "copilot-sdk-shell-rpc";

        const result = await session.rpc.shell.exec({
            command: createWriteFileCommand(markerPath, marker),
            cwd: workDir,
        });

        expect(result.processId).toBeTruthy();
        await waitForFileText(markerPath, marker);

        await session.disconnect();
    });

    it("should kill shell process", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        const command =
            os.platform() === "win32"
                ? `powershell -NoLogo -NoProfile -Command "Start-Sleep -Seconds 30"`
                : "sleep 30";

        // On Windows, terminating the shell wrapper can briefly leave grandchildren alive.
        // Keep this command outside the fixture workspace so cleanup is not blocked by cwd handles.
        const execResult = await session.rpc.shell.exec({ command, cwd: os.tmpdir() });
        expect(execResult.processId).toBeTruthy();

        const killResult = await session.rpc.shell.kill({ processId: execResult.processId });
        expect(killResult.killed).toBe(true);

        await session.disconnect();
    });

    it("should start fleet and complete custom tool task", { timeout: 180_000 }, async () => {
        const markerPath = path.join(
            workDir,
            `fleet-rpc-${Date.now()}-${Math.random().toString(36).slice(2)}.txt`
        );
        const marker = "copilot-sdk-fleet-rpc";
        const toolName = "record_fleet_completion";

        const recordFleetCompletion = defineTool(toolName, {
            description: "Records completion of the fleet validation task.",
            parameters: z.object({ content: z.string() }),
            handler: ({ content }) => {
                fs.writeFileSync(markerPath, content);
                return content;
            },
        });

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            tools: [recordFleetCompletion],
        });

        const prompt = `Use the ${toolName} tool with content '${marker}', then report that the fleet task is complete.`;

        const result = await session.rpc.fleet.start({ prompt });
        expect(result.started).toBe(true);

        await waitForFileText(markerPath, marker);

        const messages = await waitForMessages(session, (events) =>
            events.some(
                (e) =>
                    e.type === "assistant.message" &&
                    (e.data.content ?? "").toLowerCase().includes("fleet task")
            )
        );

        const userMessages = messages.filter((m) => m.type === "user.message");
        expect(userMessages.some((m) => m.data.content.includes(prompt))).toBe(true);

        const toolStarts = messages.filter((m) => m.type === "tool.execution_start");
        expect(toolStarts.some((m) => m.data.toolName === toolName)).toBe(true);

        const toolCompletes = messages.filter((m) => m.type === "tool.execution_complete");
        expect(
            toolCompletes.some(
                (m) =>
                    m.data.success === true &&
                    typeof m.data.result?.content === "string" &&
                    m.data.result.content.includes(marker)
            )
        ).toBe(true);

        const assistantMessages = messages.filter((m) => m.type === "assistant.message");
        expect(
            assistantMessages.some((m) =>
                (m.data.content ?? "").toLowerCase().includes("fleet task")
            )
        ).toBe(true);

        await session.disconnect();
    });
});
